// @sentry/react-router/build/esm/vite/makeCustomSentryVitePlugins.js
import { sentryVitePlugin } from '@sentry/vite-plugin';

/**
 * Create a custom subset of sentry's vite plugins
 */
async function makeCustomSentryVitePlugins(options) {
  const { debug, unstable_sentryVitePluginOptions, bundleSizeOptimizations, authToken, org, project, telemetry } =
    options;

  const sentryVitePlugins = sentryVitePlugin({
    authToken: authToken ?? process.env.SENTRY_AUTH_TOKEN,
    bundleSizeOptimizations,
    debug: debug ?? false,
    org: org ?? process.env.SENTRY_ORG,
    project: project ?? process.env.SENTRY_PROJECT,
    telemetry: telemetry ?? true,
    _metaOptions: {
      telemetry: {
        metaFramework: 'react-router',
      },
    },
    // will be handled in buildEnd hook
    sourcemaps: {
      disable: true,
    },
    ...unstable_sentryVitePluginOptions,
  }) ;

  // only use a subset of the plugins as all upload and file deletion tasks will be handled in the buildEnd hook
  return [
    ...sentryVitePlugins.filter(plugin => {
      return ['sentry-telemetry-plugin', 'sentry-vite-release-injection-plugin'].includes(plugin.name);
    }),
  ];
}

export { makeCustomSentryVitePlugins };
